crates/red_knot_python_semantic/resources/mdtest/narrow/conditionals/boolean.md

```diff
@@ -280,3 +280,36 @@ if x is None:
 if y is not x:
     reveal_type(y)  # revealed: str | None
 ```
+
+## `bool(..)` builtin containing expression narrows for expression
+
+```py
+def flag() -> bool: ...
+
+x = 1 if flag() else None
+
+# valid invocation, positive
+reveal_type(x)  # revealed: Literal[1] | None
+if bool(x is not None):
+    reveal_type(x)  # revealed: Literal[1]
+
+# valid invocation, negative
+reveal_type(x)  # revealed: Literal[1] | None
+if not bool(x is not None):
+    reveal_type(x)  # revealed: None
+
+# no args/narrowing
+reveal_type(x)  # revealed: Literal[1] | None
+if bool():
+    reveal_type(x)  # revealed: Literal[1] | None
```

Comment:
`bool()` is `False`, which makes the body of this unreachable. Existing type checkers do not issue diagnostics in unreachable code blocks (i.e. would not yield any `reveal_type` results here).

Maybe we could change it to `if not bool()`, but also, I'm not really sure if this test adds a lot of value. Like what would be a reasonable scenario (code change) where this would fail? So maybe we can just remove it.

Reply:
Mostly thinking about a scenario in which we assume `bool` is called with at least one argument, and do something like `&expr_call.arguments.args[0]` without checking that `.len() >= 1`

Reply:
But happy to remove if we think it's redundant or not useful code coverage!

Reply:
> Mostly thinking about a scenario in which we assume `bool` is called with at least one argument, and do something like `&expr_call.arguments.args[0]` without checking that `.len() >= 1`

That leads to a panic in the Rust code, but it would not lead to a test failure here (even if that panic would not show up for some reason).

Anyway, I'm okay with the `not bool()` test now. Thanks for the update.